src/contracts/core/PaymentCoordinator.sol

```diff
@@ -0,0 +1,502 @@
+            })
+        );
+        currPaymentCalculationEndTimestamp = paymentCalculationEndTimestamp;
+        emit DistributionRootSubmitted(rootIndex, root, paymentCalculationEndTimestamp, activatedAt);
+    }
+
+    /**
+     * @notice Create a RangePayment. Called from both `payForRange` and `payAllForRange`
+     */
+    function _payForRange(RangePayment calldata rangePayment) internal {
+        require(rangePayment.strategiesAndMultipliers.length > 0, "PaymentCoordinator._payForRange: no strategies set");
+        require(rangePayment.amount > 0, "PaymentCoordinator._payForRange: amount cannot be 0");
+        require(
+            rangePayment.duration <= MAX_PAYMENT_DURATION,
+            "PaymentCoordinator._payForRange: duration exceeds MAX_PAYMENT_DURATION"
+        );
+        require(
+            rangePayment.duration % calculationIntervalSeconds == 0,
+            "PaymentCoordinator._payForRange: duration must be a multiple of calculationIntervalSeconds"
+        );
+        require(
+            rangePayment.startTimestamp % calculationIntervalSeconds == 0,
+            "PaymentCoordinator._payForRange: startTimestamp must be a multiple of calculationIntervalSeconds"
+        );
+        require(
+            block.timestamp - MAX_RETROACTIVE_LENGTH <= rangePayment.startTimestamp &&
+                GENESIS_PAYMENT_TIMESTAMP <= rangePayment.startTimestamp,
+            "PaymentCoordinator._payForRange: startTimestamp too far in the past"
+        );
+        require(
+            rangePayment.startTimestamp <= block.timestamp + MAX_FUTURE_LENGTH,
+            "PaymentCoordinator._payForRange: startTimestamp too far in the future"
+        );
+
+        // Require rangePayment is for whitelisted strategy or beaconChainETHStrategy
+        address currAddress = address(0);
+        for (uint256 j = 0; j < rangePayment.strategiesAndMultipliers.length; j++) {
+            IStrategy strategy = rangePayment.strategiesAndMultipliers[j].strategy;
+            require(
+                strategyManager.strategyIsWhitelistedForDeposit(strategy) ||
+                    strategy == delegationManager.beaconChainETHStrategy(),
+                "PaymentCoordinator._payForRange: invalid strategy considered"
+            );
+            require(
+                currAddress < address(strategy),
+                "PaymentCoordinator._payForRange: strategies must be in ascending order to handle duplicates"
+            );
+            currAddress = address(strategy);
+        }
+
+        rangePayment.token.safeTransferFrom(msg.sender, address(this), rangePayment.amount);
+    }
+
+    /**
+     * @notice Process a payment claim.
+     * @param claim PaymentMerkleClaim struct containing the claim details for proof verification
+     * and payment transfer.
+     * @dev only callable by claimerFor[claim.earner]
+     */
+    function _processClaim(PaymentMerkleClaim calldata claim) internal onlyWhenNotPaused(PAUSED_CLAIM_PAYMENTS) {
```

Comment:
Not a fan of `internal` methods with a modifier. Could we move all this up to the external method?

(I assume there were two external methods that called this at some point?)

Reply:
I second the modifier comment

Reply:
[6f3d65e](https://github.com/Layr-Labs/eigenlayer-contracts/pull/514/commits/6f3d65e2dc454dcc0a3a57451e003455a69fa0fb)